File: java/src/main/java/com/nedap/retail/api/v1/tester/ApiWrapper.java
package com.nedap.retail.api.v1.tester;

import com.google.gson.Gson;
import com.nedap.retail.api.v1.model.*;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.net.HttpURLConnection;
import java.net.URL;
        
/**
 * A wrapper for easy access to API calls of !D Top and !D Hand
 */
public class ApiWrapper {
    private String baseUrl;
    
    /**
     * Constructor
     * @param baseUrl base URL for all API calls
     */
    public ApiWrapper(final String baseUrl) {
        // make sure baseUrl does not end with a slash
        if (baseUrl.endsWith("/")) {
            this.baseUrl = baseUrl.substring(0, baseUrl.length()-1);
        } else {
            this.baseUrl = baseUrl;
        }
    }
    
    public Status getStatus() throws Exception {
        return new Gson().fromJson(doHttpRequest("/status"), Status.class);
    }

    public SpecList getSpecs() throws Exception {
        return new Gson().fromJson(doHttpRequest("/service/events/specs"), SpecList.class);
    }
    
    public Spec createSpec(final Spec spec) throws Exception {
        return new Gson().fromJson(doHttpRequest("/service/events/specs", "POST", spec), Spec.class);
    }

    public Spec getSpec(final Integer id) throws Exception {
        return new Gson().fromJson(doHttpRequest("/service/events/specs/" + id), Spec.class);
    }

    public void deleteSpec(final Integer id) throws Exception {
        doHttpRequest("/service/events/specs/" + id, "DELETE");
    }

    public Spec updateSpec(final Spec spec) throws Exception {
        return new Gson().fromJson(doHttpRequest("/service/events/specs/" + spec.getId(), "PUT", spec), Spec.class);
    }

    public SubscriptionList getSubscriptions() throws Exception {
        return new Gson().fromJson(doHttpRequest("/service/events/subscriptions"), SubscriptionList.class);
    }
    
    public Subscription createSubscription(final Subscription subscription) throws Exception {
        return new Gson().fromJson(
                doHttpRequest("/service/events/subscriptions", "POST", subscription), Subscription.class
        );
    }

    public Subscription getSubscription(final Integer id) throws Exception {
        return new Gson().fromJson(doHttpRequest("/service/events/subscriptions/" + id), Subscription.class);
    }

    public void deleteSubscription(final Integer id) throws Exception {
        doHttpRequest("/service/events/subscriptions/" + id, "DELETE");
    }

    public Subscription updateSubscription(final Subscription subscription) throws Exception {
        return new Gson().fromJson(
                doHttpRequest(
                        "/service/events/subscriptions/" + subscription.getId(), "PUT", subscription
                ), Subscription.class
        );
    }

    public void sendActions(final Actions actions) throws Exception {
        doHttpRequest("/service/actions", "POST", actions);
    }

    public Settings getSettings() throws Exception {
        return new Gson().fromJson(doHttpRequest("/service/settings"), Settings.class);
    }
    
    public void updateSettings(final Settings settings) throws Exception {
        doHttpRequest("/service/settings", "PUT", settings);
    }

    public void heartbeat() throws Exception {
        doHttpRequest("/heartbeat", "GET");
    }

    public void testConnection() throws Exception {
        final String httpResult = doHttpRequest("/status");
        if (httpResult.length() > 4) {
            System.out.println("Connection OK");
        } else {
            System.out.println("Invalid response received");
        }
    }

    private String doHttpRequest(final String url, final String requestMethod, final Object data) throws Exception {
        final HttpURLConnection connection = (HttpURLConnection) new URL(this.baseUrl + url).openConnection();
        connection.setConnectTimeout(10000);
        connection.setReadTimeout(10000);
        if (requestMethod.equals("POST") || requestMethod.equals("PUT")) {
            connection.setDoOutput(true);
            connection.addRequestProperty("Content-Type", "application/json");
            connection.setRequestMethod(requestMethod);
            try (final OutputStreamWriter out = new OutputStreamWriter(connection.getOutputStream())) {
                out.write(new Gson().toJson(data));
            }
        } else if (requestMethod.equals("DELETE")) {
            connection.setDoOutput(true);
            connection.addRequestProperty("Content-Type", "application/json");
            connection.setRequestMethod(requestMethod);
            connection.connect();
        }
        System.out.println("Response code = " + connection.getResponseCode());
        try (final BufferedReader in = new BufferedReader(new InputStreamReader(connection.getInputStream()))) {
            final StringBuilder result = new StringBuilder();
            String line;
            while ((line = in.readLine()) != null) {
                result.append(line);
            }
            return result.toString();
        }
    }
    
    private String doHttpRequest(final String url, final String requestMethod) throws Exception {
        return doHttpRequest(url, requestMethod, null);
    }

    private String doHttpRequest(final String url) throws Exception {
        return doHttpRequest(url, "GET", null);
    }
}

File: csharp/ApiWrapper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Net;
using System.IO;
using System.Runtime.Serialization.Json;

namespace StoreIDTopGateAPI
{
    class ApiWrapper
    {
        private String baseUrl;

        /**
         * Constructor
         * @param url base URL for all API calls
         */
        public ApiWrapper(String baseUrl)
        {
            // make sure baseUrl does not end with a slash
            if (baseUrl.EndsWith("/"))
            {
                this.baseUrl = baseUrl.Substring(0, baseUrl.Length - 1);
            }
            else
            {
                this.baseUrl = baseUrl;
            }
        }

        public Status getStatus()
        {
            String httpResult = doHttpRequest("/status");
            return (Status)fromJson(httpResult, typeof(Status));
        }

        public SpecList getSpecs()
        {
            String httpResult = doHttpRequest("/service/events/specs");
            return (SpecList)fromJson(httpResult, typeof(SpecList));
        }
        
        public Spec createSpec(Spec spec)
        {
            String httpResult = doHttpRequest("/service/events/specs", "POST", toJson(spec, typeof(Spec)));
            return (Spec)fromJson(httpResult, typeof(Spec));
        }
        
        public Spec getSpec(int id)
        {
            String httpResult = doHttpRequest("/service/events/specs/" + id);
            return (Spec)fromJson(httpResult, typeof(Spec));
        }
        
        public void deleteSpec(int id)
        {
            String httpResult = doHttpRequest("/service/events/specs/" + id, "DELETE");
        }
        
        public Spec updateSpec(Spec spec)
        {
            String httpResult = doHttpRequest("/service/events/specs/" + spec.id, "PUT", toJson(spec, typeof(Spec)));
            return (Spec)fromJson(httpResult, typeof(Spec));
        }
    
        public SubscriptionList getSubscriptions()
        {
            String httpResult = doHttpRequest("/service/events/subscriptions");
            return (SubscriptionList)fromJson(httpResult, typeof(SubscriptionList));
        }
        
        public Subscription createSubscription(Subscription subscription)
        {
            String httpResult = doHttpRequest("/service/events/subscriptions", "POST", toJson(subscription, typeof(Subscription)));
            return (Subscription)fromJson(httpResult, typeof(Subscription));
        }
        
        public Subscription getSubscription(int id)
        {
            String httpResult = doHttpRequest("/service/events/subscriptions/" + id);
            return (Subscription)fromJson(httpResult, typeof(Subscription));
        }
        
        public void deleteSubscription(int id)
        {
            String httpResult = doHttpRequest("/service/events/subscriptions/" + id, "DELETE");
        }
        
        public Subscription updateSubscription(Subscription subscription) {
            String httpResult = doHttpRequest("/service/events/subscriptions/" + subscription.id, "PUT", toJson(subscription, typeof(Subscription)));
            return (Subscription)fromJson(httpResult, typeof(Subscription));
        }

        public void sendActions(Actions actions)
        {
            String httpResult = doHttpRequest("/service/actions", "POST", toJson(actions, typeof(Actions)));
        }

        public Settings getSettings()
        {
            String httpResult = doHttpRequest("/service/settings", "GET");
            return (Settings)fromJson(httpResult, typeof(Settings));
        }

        public void updateSettings(Settings settings)
        {
            String httpResult = doHttpRequest("/service/settings", "PUT", toJson(settings, typeof(Settings)));
        }

        public void heartbeat()
        {
            String httpResult = doHttpRequest("/heartbeat", "GET");
        }

        public void testConnection() {
            try {
                String httpResult = doHttpRequest("/status");
                if (httpResult.Length>4)
                {
                    Console.WriteLine("Connection OK");
                }
                else
                {
                    Console.WriteLine("Invalid response received");
                }
            }
            catch (Exception e)
            {
                Console.WriteLine(e.ToString());
            }
        }

        /**
         * Helper function for HTTP requests
         */
        private String doHttpRequest(String url, String requestMethod = "GET", String data = "")
        {
            byte[] dataBytes = System.Text.Encoding.UTF8.GetBytes(data);
            HttpWebRequest request = (HttpWebRequest)HttpWebRequest.Create(this.baseUrl + url);
            if (requestMethod == "GET")
            {
                // is the default, do nothing
            }
            else if ((requestMethod == "POST") || (requestMethod == "PUT"))
            {
                request.Method = requestMethod;
                request.ContentType = "application/json";
                request.ContentLength = dataBytes.Length;
                Stream dataStream = request.GetRequestStream();
                dataStream.Write(dataBytes, 0, dataBytes.Length);
                dataStream.Close();
            }
            else if (requestMethod == "DELETE")
            {
                request.Method = requestMethod;
            }
            HttpWebResponse response = (HttpWebResponse)request.GetResponse();
            Console.WriteLine("Response code = " + response.StatusCode);
            StreamReader streamReader = new StreamReader(response.GetResponseStream());
            String result = "";

            while(streamReader.Peek() != -1)
            {
                result += streamReader.ReadLine();
            }
            return result;
        }

        /**
         * Helper function for JSON serialization
         * @see http://msdn.microsoft.com/en-us/library/bb412179.aspx
         */
        private Object fromJson(String json, Type theClass)
        {
            DataContractJsonSerializer ser = new DataContractJsonSerializer(theClass);
            byte[] dataBytes = System.Text.Encoding.UTF8.GetBytes(json);
            MemoryStream stream = new MemoryStream(dataBytes);
            return ser.ReadObject(stream);
        }

        /**
         * Helper function for JSON serialization
         * @see http://msdn.microsoft.com/en-us/library/bb412179.aspx
         */
        private String toJson(Object obj, Type theClass)
        {
            DataContractJsonSerializer ser = new DataContractJsonSerializer(theClass);
            MemoryStream stream = new MemoryStream();
            ser.WriteObject(stream, obj);
            stream.Position = 0;
            StreamReader streamreader = new StreamReader(stream);
            return streamreader.ReadToEnd();
        }
    }
}
